### PR TITLE
Remove TokenTimelock, PaymentSplitter, ERC20Snapshot, ERC20VotesComp, GovernorVotesComp

### PR DESCRIPTION
**Pull request by Amxx**
Fixes LIB-874

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changeset entry (run `npx changeset add`)

### CODE REVIEW

**Comment by frangio** [comment]
Upgradeable patch needs to be updated.

File: .changeset/clever-eagles-unite.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'openzeppelin-solidity': major
+---
+
+`PaymentSplitter`: removed.
```
Comment (frangio) [inline .changeset/clever-eagles-unite.md:5]: This is not a great changeset. We had a similar unresolved conversation with the previous round of deletions.

Maybe we should first think about what we want the changelog to look like, and from there we should figure out what changesets we need...

What I imagine is a single section called "Removals" with a section for each of the removal reasons. If so, I think we can have one changeset for each of those sections.

File: contracts/finance/VestingWallet.sol [hunk 2]
```diff
@@ -136,7 +146,7 @@ contract VestingWallet is Context {
     function _vestingSchedule(uint256 totalAllocation, uint64 timestamp) internal view virtual returns (uint256) {
         if (timestamp < start()) {
             return 0;
-        } else if (timestamp > start() + duration()) {
+        } else if (timestamp > end()) {
```
[the remark this answers is not in the document]
Comment (Amxx) [reply contracts/finance/VestingWallet.sol:149]: I have. If we do so, should we keep `end()` virtual ? I'm thinking yes, but I don't have very strong arguments

File: .changeset/clever-eagles-unite.md (hunk 1 above)
Comment (Amxx) [reply .changeset/clever-eagles-unite.md:5]: Yes, I'll start that section in this PR, and we can move changesets from previous PR to that section in a dedicated PR.

File: contracts/finance/VestingWallet.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (frangio) [reply contracts/finance/VestingWallet.sol:149]: I'm fine with making this virtual in this PR, as we've done until now according to our guidelines, but I would love to review that. In the tokens, we've already decided to make the `_mint`, `_transfer`, etc., "aliases" that are non-virtual functions, but this one is not exactly an alias in the same way. So maybe let's discuss this separately.

**Review by frangio — APPROVED** [review]

**Comment by ItsCuzzo** [comment]
Hey there, out of curiosity what was the rationale behind the removal of PaymentSplitter? It was quite a useful contract that I have utilised a variety of times during development, particularly with ERC-721 token contracts with multiple shareholders.

**Comment by Amxx** [comment]
Hello @ItsCuzzo 

The PaymentSplitter received a lot of criticism related to the inability to update the share. Some people would expect the shares to be transferable using an ERC20 or ERC721 interface. 

The plan is to remove the old version (that suffers limitations) in version 5.0, and then reintroduce a better version down the line in 5.1 or 5.2. In the meantime, the old version remains available in the 4.X versions.

**Comment by ItsCuzzo** [comment]
Awesome, I did see some of that discussion now that you mention it. Thanks for the detailed reply!

**Comment by Arvolear** [comment]
Hey, is it the same story with `ERC20Snapshot`? It was indeed ultra useful for some lightweight DAOs.

**Comment by frangio** [comment]
Please say more about "ultra lightweight"?

**Comment by Arvolear** [comment]
I mean DAOs with `ERC20Snapshot` as a voting token.

**Comment by Amxx** [comment]
`ERC20Snapshot` does not support delegation, and requires someone to create a snapshot when needed (part of the proposal workflow I guess).

This doesn't sound like the best design.

**Comment by Arvolear** [comment]
There are many approaches to build DAOs and I agree that the absence of delegation might break some design decision.

But there is one case with 2 step DAOs. In these DAOs there is a set of validators that is elected by regular members to veto proposals.

These validators should not delegate and they have fixed voting power. Snapshots here fix the problem when there are 2 proposals (regular and to remove a validator). 

Let's say the regular one lasts for 2 weeks and the removal is only 1 week long.

So if validator gets removed, he won't be able to vote on the regular proposal even though it was created first.

**Comment by Amxx** [comment]
To add more context:

- `ERC20Snapshot` was released before `ERC20Votes`, it was a first approach.
- When `ERC20Votes` was released it was considered the better option. In particular becauseits a more standard interface that is supported by some third party tools (such as Tally)
- Our governor is compatible with `ERC20Votes` (and `ERC721Votes`), but not with `ERC20Snapshot`.

Planning for 5.0, we felt that the `Votes` module was the go to solution, and that `ERC20Snapshot` was no longer relevant. We decided to deprecate it, and it is now removed for the repo. It remains available in the 4.x packages.

If you really want, you should be able to  adapt `ERC20Snapshot` to 5.0 easily.

**Comment by ernestognw** [comment]
We got similar questions quite a few times and I think we need to provide a bit more clarity on how to replace ERC20Snapshot. Although I agree with @Amxx in that people should generally use `ERC20Votes` for governance (idk how the "lightweight" DAO case works), it's true there are use cases for snapshotting that aren't related to governance and the alternatives aren't trivial:

- On one hand, you can use the 4.x version of ERC20Snapshot, but if you mix it with 5.x contracts, they wont be guaranteed to have compatible storage and semantics with 4.x. In a new minor version, we may add new storage variables or slightly modify the semantics of a function in a way that mixed with a 4.x's ERC20Snapshot could break some of your assumptions.
- On the other, you can use a modification of ERC20Votes to mimic the ERC20Snapshot behavior, but I don't think that's as easy as it sounds, and requires a bit of context.

I'd generally go for adapting an ERC20Votes because it guarantees support during the 5.x version, but given that the contract is built for governance purposes, I see there are a few important differences to ERC20Snapshot:

- The "snapshots" (now called checkpoints) depend on the [EIP-6372 contract](https://eips.ethereum.org/EIPS/eip-6372), as opposed to an `snapshotId`.
- The `Votes` contract counts delegated tokens ([by default to the address 0](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/932fddf69a699a9a80fd2396fd1a2ab91cdda123/contracts/governance/utils/Votes.sol#L128))

So the ERC20Votes should be overridden to use a counter as `clock()` and have self-delegation by default. [Here's a PoC](https://gist.github.com/ernestognw/a731a75de7330c472303275f21a4d468)
 for reference purposes, note its unaudited code.

Also, an ERC20Snapshot implemented from scratch is also relatively easy using the OpenZeppelin Checkpoints primitive. [Here's a reference PoC](https://gist.github.com/ernestognw/6a4a8e15e9435dc67ad17f327701e397). Unaudited as well.

**Comment by codebuster22** [comment]
Hey, if the PaymentSplitter is not being worked on, can I take it up and create a payment splitter? I can start creating specs and once approved, develop it.
Needed it for a project.
@frangio @Amxx

**Comment by Amxx** [comment]
@codebuster22 If there is no visible on the payment splitter, its because the believe there is no solution to implement the features we want to have. We are considering relaxing our requirements, and we don't now for sure in which direction to relax them (which guarantee/feature to drop). Please be patient (and possibly use the version in 4.9.x in the meantime)

**Pull request merged**